src/core/aws-net-websocket-registry.adb

```diff
@@ -551,6 +562,37 @@ package body AWS.Net.WebSocket.Registry is
          Registered.Clear;
       end Finalize;
 
+      ----------
+      -- Free --
+      ----------
+
+      procedure Free_Or_Defer (WebSocket : in out Object_Class) is
+      begin
+         --  If WebSocket is in Sending it means that it has been
+         --  taken by the Get_Socket call. We cannot free it now, we
+         --  record this socket to be freed as soon as it is released
+         --  (Release_Socket) call.
+
+         if Sending.Contains (WebSocket.Id) then
+            WebSocket.To_Free := True;
+
```

Comment:
wondering whether we should reset WebSocket to null in this case as well. Users should really not try to use it anymore after calling Free_Or_Defer anyway, so this might avoid subtle bugs (where code behaves differently depending on whether we actually freed or deferred.

Reply:
Not sure, in the other Free_Or_Defer we have:
`   Free_Or_Defer (Registered (Id));`
If you set WebSocket to null it will also nullify the pointer recorded into Registered container.

Reply:
But isn't that the point: we should first unregister and then free_or_defer. Otherwise, the risk is high that later someone will do additional changes after the call and not realized that it is unsafe to call the pointer which might have been freed.

Reply:
> But isn't that the point: we should first unregister and then free_or_defer.

No the part quoted is registering the WebSocket to be freed, but we still need the handle until we actually release the memory in the object and the pointer itself. Ot there is really something that I don't understand in your proposal.

Reply:
I am not sure I get your point. In the case where we can actually free memory, the pointer will be reset to null, so the rest of the code cannot depend on still having the pointer afterwards. The caller to Free_Or_Defer should always expect that the pointer can be reset to null. So it seems like the code `Free_Or_Defer (Registered (Id))` has wrong expectations here.

```
      procedure Free_Or_Defer (Id : UID) is
      begin
         if Registered.Contains (Id) then
            Free_Or_Defer (Registered (Id));
         end if;
      end Free_Or_Defer;
```

This code likely should unregister the socket too. If I understand the code correctly (and that's not a given... :-) the socket will be freed as soon as it is released anyway, so we no longer need to have an entry for the socket in the registry. So perhaps something like:

```
      procedure Free_Or_Defer (Id : UID) is
         W : Object_Class;
         C : WebSocket_Set.Cursor;
      begin
         C := Registered.Find (Id);
         if Has_Element (C) then
             W := Element (C);
             Registered.Delete (C);
             Free_Or_Defer (W);
         end if;
      end Free_Or_Defer;
```

I might definitely be wrong, the code is not so easy to follow.